File: src/main/java/com/polygloat/service/LanguageService.java
package com.polygloat.service;

import com.polygloat.collections.LanguageSet;
import com.polygloat.constants.Message;
import com.polygloat.dtos.request.LanguageDTO;
import com.polygloat.exceptions.NotFoundException;
import com.polygloat.model.Language;
import com.polygloat.model.Repository;
import com.polygloat.repository.LanguageRepository;
import lombok.RequiredArgsConstructor;
import lombok.Setter;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import javax.persistence.EntityManager;
import java.util.Collection;
import java.util.Optional;
import java.util.Set;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor(onConstructor = @__(@Autowired))
public class LanguageService {
    private final LanguageRepository languageRepository;
    private final EntityManager entityManager;

    @Setter(onMethod = @__(@Autowired))
    private TranslationService translationService;

    @Transactional
    public Language createLanguage(LanguageDTO dto, Repository repository) {
        Language language = Language.fromRequestDTO(dto);
        language.setRepository(repository);
        repository.getLanguages().add(language);
        languageRepository.save(language);
        return language;
    }

    @Transactional
    public void deleteLanguage(Long id) {
        Language language = languageRepository.findById(id).orElseThrow(NotFoundException::new);
        translationService.deleteAllByLanguage(language.getId());
        languageRepository.delete(language);
    }

    @Transactional
    public Language editLanguage(LanguageDTO dto) {
        Language language = languageRepository.findById(dto.getId()).orElseThrow(NotFoundException::new);
        language.updateByDTO(dto);
        entityManager.persist(language);
        return language;
    }

    public LanguageSet getImplicitLanguages(Repository repository) {
        return repository.getLanguages().stream().limit(2).collect(Collectors.toCollection(LanguageSet::new));
    }

    @Transactional
    public LanguageSet findAll(Long repositoryId) {
        return new LanguageSet(languageRepository.findAllByRepositoryId(repositoryId));
    }

    public Optional<Language> findById(Long id) {
        return languageRepository.findById(id);
    }

    public Optional<Language> findByAbbreviation(String abbreviation, Repository repository) {
        return languageRepository.findByAbbreviationAndRepository(abbreviation, repository);
    }

    public Optional<Language> findByAbbreviation(String abbreviation, Long repositoryId) {
        return languageRepository.findByAbbreviationAndRepositoryId(abbreviation, repositoryId);
    }

    public LanguageSet findByAbbreviations(Collection<String> abbreviations, Long repositoryId) {
        Set<Language> langs = languageRepository.findAllByAbbreviationInAndRepositoryId(abbreviations, repositoryId);
        if (!langs.stream().map(Language::getAbbreviation).collect(Collectors.toSet()).containsAll(abbreviations)) {
            throw new NotFoundException(Message.LANGUAGE_NOT_FOUND);
        }
        return new LanguageSet(langs);
    }

    @Transactional
    public Language getOrCreate(Repository repository, String languageAbbreviation) {
        return this.findByAbbreviation(languageAbbreviation, repository)
                .orElseGet(
                        () -> this.createLanguage(
                                LanguageDTO.builder()
                                        .abbreviation(languageAbbreviation)
                                        .name(languageAbbreviation).build(),
                                repository)
                );
    }

    public LanguageSet getLanguagesForTranslationsView(Set<String> languages, Repository repository) {
        if (languages == null) {
            return getImplicitLanguages(repository);
        }
        return findByAbbreviations(languages, repository.getId());
    }

    public Optional<Language> findByName(String name, Repository repository) {
        return languageRepository.findByNameAndRepository(name, repository);
    }

    public void deleteAllByRepository(Long repositoryId) {
        languageRepository.deleteAllByRepositoryId(repositoryId);
    }
}

File: src/main/java/io/tolgee/service/LanguageService.kt
package io.tolgee.service

import io.tolgee.collections.LanguageSet
import io.tolgee.constants.Message
import io.tolgee.dtos.request.LanguageDto
import io.tolgee.exceptions.NotFoundException
import io.tolgee.model.Language
import io.tolgee.model.Language.Companion.fromRequestDTO
import io.tolgee.model.Project
import io.tolgee.repository.LanguageRepository
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.data.domain.Page
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.Pageable
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.util.*
import java.util.stream.Collectors
import javax.persistence.EntityManager

@Service
class LanguageService(
        private val languageRepository: LanguageRepository,
        private val entityManager: EntityManager,
) {
    private var translationService: TranslationService? = null

    @Transactional
    fun createLanguage(dto: LanguageDto?, project: Project): Language {
        val language = fromRequestDTO(dto!!)
        language.project = project
        project.languages.add(language)
        languageRepository.save(language)
        return language
    }

    @Transactional
    fun deleteLanguage(id: Long) {
        val language = languageRepository.findById(id).orElseThrow { NotFoundException() }
        translationService!!.deleteAllByLanguage(language.id)
        languageRepository.delete(language)
    }

    @Transactional
    fun editLanguage(id: Long, dto: LanguageDto): Language {
        val language = languageRepository.findById(id).orElseThrow { NotFoundException() }
        language.updateByDTO(dto)
        entityManager.persist(language)
        return language
    }

    fun getImplicitLanguages(project: Project): LanguageSet {
        val data = getPaged(projectId = project.id, PageRequest.of(0, 2))
        return data.content.toCollection(LanguageSet())
    }

    @Transactional
    fun findAll(projectId: Long?): LanguageSet {
        return LanguageSet(languageRepository.findAllByProjectId(projectId))
    }

    fun findById(id: Long): Optional<Language> {
        return languageRepository.findById(id)
    }

    fun findByTag(tag: String, project: Project): Optional<Language> {
        return languageRepository.findByTagAndProject(tag, project)
    }

    fun findByTag(tag: String?, projectId: Long): Optional<Language> {
        return languageRepository.findByTagAndProjectId(tag, projectId)
    }

    fun findByTags(tag: Collection<String>?, projectId: Long?): LanguageSet {
        val langs = languageRepository.findAllByTagInAndProjectId(tag, projectId)
        if (!langs.stream().map(Language::tag).collect(Collectors.toSet()).containsAll(tag!!)) {
            throw NotFoundException(Message.LANGUAGE_NOT_FOUND)
        }
        return LanguageSet(langs)
    }

    fun getLanguagesForTranslationsView(languages: Set<String>?, project: Project): LanguageSet {
        return if (languages == null) {
            getImplicitLanguages(project)
        } else findByTags(languages, project.id)
    }

    fun findByName(name: String?, project: Project): Optional<Language> {
        return languageRepository.findByNameAndProject(name, project)
    }

    fun deleteAllByProject(projectId: Long?) {
        languageRepository.deleteAllByProjectId(projectId)
    }

    @Autowired
    fun setTranslationService(translationService: TranslationService?) {
        this.translationService = translationService
    }

    fun saveAll(languages: Iterable<Language>): MutableList<Language>? {
        return this.languageRepository.saveAll(languages)
    }

    fun getPaged(projectId: Long, pageable: Pageable): Page<Language> {
        return this.languageRepository.findAllByProjectId(projectId, pageable)
    }
}
